src/d3d9/d3d9_interop.cpp

```diff
@@ -117,6 +118,86 @@ namespace dxvk {
     return S_OK;
   }
 
+  ////////////////////////////////
+  // Buffer Interop
+  ///////////////////////////////
+
+  D3D9VkInteropBuffer::D3D9VkInteropBuffer(
+          IUnknown*             pInterface,
+          D3D9CommonBuffer*     pBuffer)
+    : m_interface(pInterface)
+    , m_buffer  (pBuffer)
+    , m_texture (nullptr) {
+  }
+
+  D3D9VkInteropBuffer::D3D9VkInteropBuffer(
+    IUnknown* pInterface,
+    D3D9CommonTexture* pTexture)
+    : m_interface(pInterface)
+    , m_buffer(nullptr)
+    , m_texture(pTexture) {
+  }
+
+  D3D9VkInteropBuffer::~D3D9VkInteropBuffer() {
+
+  }
+
+  ULONG STDMETHODCALLTYPE D3D9VkInteropBuffer::AddRef() {
+    return m_interface->AddRef();
+  }
+  
+  ULONG STDMETHODCALLTYPE D3D9VkInteropBuffer::Release() {
+    return m_interface->Release();
+  }
+  
+  HRESULT STDMETHODCALLTYPE D3D9VkInteropBuffer::QueryInterface(
+          REFIID                riid,
+          void**                ppvObject) {
+    return m_interface->QueryInterface(riid, ppvObject);
+  }
+
+
+  HRESULT STDMETHODCALLTYPE D3D9VkInteropBuffer::GetBufferInfo(
+          D3D9VkBufferType       Type,
+          VkBuffer*              pHandle,
+          VkBufferCreateInfo*    pInfo) {
+    Rc<DxvkBuffer> buffer;
+    if (m_buffer) {
+      switch (Type)
+      {
+        case D3D9_VK_BUFFER_TYPE_MAPPING:
+          buffer = m_buffer->GetBuffer<D3D9_COMMON_BUFFER_TYPE_MAPPING>();
+          break;
+        case D3D9_VK_BUFFER_TYPE_STAGING:
+          buffer = m_buffer->GetBuffer<D3D9_COMMON_BUFFER_TYPE_MAPPING>();
+          break;
+        case D3D9_VK_BUFFER_TYPE_REAL:
+          buffer = m_buffer->GetBuffer<D3D9_COMMON_BUFFER_TYPE_REAL>();
+          break;
+        default:
+          return D3DERR_INVALIDCALL;
+      }
+    } else if (m_texture) {
+      buffer = m_texture->GetBuffer();
+    }
+
+    if (pHandle) {
+      *pHandle = buffer->storage()->getBufferInfo().buffer;
```

Comment:
This is not safe, backing storage can change at any time either due to defrag or copy optimizations. This either needs a full sync with the CS thread with a guarantee that the interop client finishes using the buffer before submitting any more DXVK work, or lock the buffer in place by calling `DxvkContext::ensureBufferAddress` *from the CS thread* in order to prevent the backend from relocaing it (see `D3D9DeviceEx::InjectCs`).

If this is ever used to expose mapped buffers that can get discarded by the front-end for any reason, things get even more problematic.

Moreover, this is missing the buffer offset. DXVK only creates one `VkBuffer` per memory allocation these days unless there's a good reason not to (e.g. index buffers on a device that doesn't support `KHR_maintenance5`).

Reply:
> This is not safe, backing storage can change at any time either due to defrag or copy optimizations. This either needs a full sync with the CS thread with a guarantee that the interop client finishes using the buffer before submitting any more DXVK work, or lock the buffer in place by calling `DxvkContext::ensureBufferAddress` _from the CS thread_ in order to prevent the backend from relocaing it (see `D3D9DeviceEx::InjectCs`).
> 
> If this is ever used to expose mapped buffers that can get discarded by the front-end for any reason, things get even more problematic.
> 
> Moreover, this is missing the buffer offset. DXVK only creates one `VkBuffer` per memory allocation these days unless there's a good reason not to (e.g. index buffers on a device that doesn't support `KHR_maintenance5`).

The context I'm using interop in is completely manually managed in terms of thread/sharing safety. DXVK and the graphics queue are flushed and locked while submitting work, and I'm looking into the possibility of submitting work directly to the DXVK command stream as a separate (much smaller) PR. This is a very good point though and I can envision this being a footgun for parallel stuff or for clients that aren't managing their locks. I'll see if I can come up with a better solution.

Reply:
> Moreover, this is missing the buffer offset. DXVK only creates one `VkBuffer` per memory allocation these days unless there's a good reason not to (e.g. index buffers on a device that doesn't support `KHR_maintenance5`).

Gotcha